### PR TITLE
Add 3D variants of CNN networks

### LINKED ISSUE #763: General nn/layers/op file to gracefully handle 3D vs 2D
### Environment (optional)

_No response_

### Describe the feature

A collection of standard vision layers / components / ops (convs, pooling, upsampling, norms, ...) with a dim keyword that can return the 3D or 2D variant. Other option is using a dict e.g. `convnd = {2 : nn.Conv2d, 3 : nn.Conv3d}`

### Motivation

As the repo starts to tackle 3D data, various networks will need to have a 3D variant. The idea above can reduce overall code complexity and increase readability.

### Alternatives considered

_No response_

### Additional context

Related to my comment on PR #743

### PR DESCRIPTION
**Pull request by nucli-vicky**
- [x] DIP (tests: also not done for 2D)
- [x] DRUNet (tests :white_check_mark: ; but might want to add parametrized tests for the different act/downsample/upsample modes)
- [x] UNet (tests :white_check_mark:)
- [x] GAN networks (tests :white_check_mark:)
- [x] DnCNN (tests :white_check_mark:)
- [x] ICNN (tests :white_check_mark:)
- [x] DScCP (tests :white_check_mark:)
- [x] PDNet (tests :white_check_mark:)
- [x] Add or parametrize tests so we have coverage for 3D variants

Will add the 3D variants for diffusion and nets using attention at a different time, as I want to do proper testing with some training runs for it.

This PR will not make `test_onesplit` support 3D. This will be fixed when sliding window inference is added together with a refactor of `PatchDataset`.

This PR should fix #763 and is related to #804 and #743 

**Notes:**
- Also fixed some docs, typing, and added flexibility (e.g. `ESRGANDiscriminator` now has a `filters` argument)
- Most notable: Tests now check all the possible downsampling, upsampling, and activation functions for `DRUNet`. After I wrote the test, it actually failed, and main failed too, when we specify maxpooling. This bug originated from the fact that the default `padding` in `downsample_maxpool` was 0 and is now fixed.

### Checks to be done before submitting your PR

- [x] `python3 -m pytest deepinv/tests` runs successfully.
- [x] `black .` runs successfully.
- [x] `make html` runs successfully (in the `docs/` directory).
- [x] Updated docstrings related to the changes (as applicable).
- [x] Added an entry to the [CHANGELOG.rst](../CHANGELOG.rst).

### CODE REVIEW

**Comment by nucli-vicky** [comment]
While I am looking through the tests, I saw there is a `if __name__ == __main__:` randomly inside `tests/dummy.py` which simply plots a sample from `DummyCircles`. Can this be removed as it's not adding anything to the code?
Also, the following line in `DummyCircles` does not make sense to me:

`max_rad = max(imsize[0], imsize[1]) / 2`

because `imsize` is format (C, H, W). Wouldn't 

`max_rad = min(imsize[1], imsize[2]) / 2`

make more sense?

**Comment by Tmodrzyk** [comment]
Thanks for this very nice PR Vicky, I'll be reviewing it shortly

**Comment by Andrewwango** [comment]
@nucli-vicky how does the 3D DRUnet contribution overlap/duplicate #743 ? If it removes the need to do further work on #743 then there's less work for @chaithyagr

File: deepinv/models/dip.py [hunk 1]
```diff
@@ -41,36 +62,34 @@ def __init__(self, img_size, in_size=(4, 4), layers=7, channels=256):
...
             self.net.add(nn.Upsample(size=hidden_size[i], mode="nearest"))
-            conv = nn.Conv2d(
+            c = conv(
                 channels,
                 channels,
                 kernel_size,
                 strides[i],
                 padding=(kernel_size - 1) // 2,
                 bias=True,
             )
```
Comment (Tmodrzyk) [inline deepinv/models/dip.py]: I think the linter will soon complain about one letter variables, maybe we could change it to ``conv_layer`` ?

File: deepinv/models/dncnn.py [hunk 2]
```diff
@@ -116,10 +126,3 @@ def weights_init_kaiming(m):
     classname = m.__class__.__name__
     if classname.find("Conv") != -1:
         nn.init.kaiming_normal_(m.weight.data, a=0, mode="fan_in")
-    elif classname.find("Linear") != -1:
-        nn.init.kaiming_normal_(m.weight.data, a=0, mode="fan_in")
-    elif classname.find("BatchNorm") != -1:
-        m.weight.data.normal_(mean=0, std=math.sqrt(2.0 / 9.0 / 64.0)).clamp_(
-            -0.025, 0.025
-        )
-        nn.init.constant(m.bias.data, 0.0)
```
Comment (Tmodrzyk) [inline deepinv/models/dncnn.py:125]: I'm not sure why this was necessary in the first place, but why is it not anymore ?

File: deepinv/tests/test_adversarial.py [hunk 3]
```diff
@@ -100,3 +105,32 @@ def test_adversarial_training(combo_name, imsize, device, physics, dataset):
...
+        img_size=test_shape, filters=(8, 8, 8), dim=dim
+    )
+    batch = torch.randn([1] + list(test_shape))
+    out = discriminator(batch)
+    assert out.shape == tuple([1] + list(discriminator.output_shape))
+
+    discriminator = dinv.models.PatchGANDiscriminator(ndf=2, dim=dim)
+    batch = torch.randn([1] + list(test_shape))
+    out = discriminator(batch)
+    assert out.shape == tuple([1, 1] + [11 for i in range(dim)])
```
Comment (Tmodrzyk) [inline deepinv/tests/test_adversarial.py:136]: Thanks, this should have been here for a long time !

**Review by Tmodrzyk — CHANGES_REQUESTED** [review]
Thanks for this awesome PR @nucli-vicky. Also thank you for all the type annotations / new tests. I have a few comments / questions but overall LGTM, I think this can safely be merged.
Edit: as Andrew mentioned, before approving I'd need to know if this PR totally / partially implements #743 ?

**Comment by nucli-vicky** [comment]
@Andrewwango @Tmodrzyk This PR partially implements the changes needed for #743 . However, #743 also requires some work on handling complex data types and a small change to `dpir.py`. 
I like how #743 handles the complex input / output using `MRIMixin`, and think it might be nice to bring it in a general form. In principle we could make it for any such image-to-image network through inheritance.

As an example; with `DRUNet`:
```python
class DRUNetMRIMixin(DRUNet):
   def forward(self, x, sigma):
       x = MRIMixin.from_torch_complex(x)
       super().forward(x, sigma)
       return MRIMixin.to_torch_complex(x)
```

We could make a function that takes a network class and returns such a new one for use with `MRIMixin` as defined above.

I think it's best if #743 continues by building from this PR and making this inheritance style function so that all CNNs with 3D support can now be used for MRI :), and ofcourse finishing the other things which already started in #743 that are unrelated to 3D models specifically.

File: deepinv/models/dncnn.py (hunk 2 above)
Comment (nucli-vicky) [reply deepinv/models/dncnn.py:125]: It was not necessary in the first place, and never used. I believe it originates from some copy paste. Since it is used nowhere, I deleted it.

File: deepinv/models/dip.py (hunk 1 above)
Comment (nucli-vicky) [reply deepinv/models/dip.py]: Good point, thanks!

File: deepinv/tests/test_models.py [hunk 4]
```diff
@@ -507,6 +507,34 @@ def test_denoiser_sigma_color(batch_size, denoiser, device):
...
+}
+
+
+@pytest.mark.parametrize("model", MODELS_3D.items())
+def test_3d_nets(model, device):
+    default_args = {"in_channels": 1, "out_channels": 1, "dim": "3d", "device": device}
+    model, args = model
+    args.update(default_args)
+    net = model(**args)
+    test_tensor = torch.randn((1, 1, 32, 32, 32), device=device)
```
Comment (Andrewwango) [inline deepinv/tests/test_models.py:533]: Maybe it would be good to test with some esoteric size like that in the imsize fixture

File: deepinv/tests/test_models.py [hunk 5]
```diff
@@ -507,6 +507,34 @@ def test_denoiser_sigma_color(batch_size, denoiser, device):
...
+        "nc": [8, 8, 8, 8],
+        "nb": 2,
+        "pretrained": None,
+    },
+    dinv.models.UNet: {"scales": 2},
+}
+
+
+@pytest.mark.parametrize("model", MODELS_3D.items())
+def test_3d_nets(model, device):
```
Comment (Andrewwango) [inline deepinv/tests/test_models.py]: This is more of a 3d denoiser test, no?

File: deepinv/tests/test_models.py [hunk 6]
```diff
@@ -507,6 +507,34 @@ def test_denoiser_sigma_color(batch_size, denoiser, device):
         assert x_hat.shape == x.shape
 
 
+MODELS_3D = {
```
Comment (Andrewwango) [inline deepinv/tests/test_models.py]: I was thinking maybe we should advertise somewhere in the user guide that lots of denoiser accept 3D data?

File: deepinv/models/unet.py [hunk 7]
```diff
@@ -71,36 +73,52 @@ class UNet(Denoiser):
...
+        batch_norm: bool | str = True,
+        scales: int = 4,
+        device: torch.device | str = None,
+        dim: str | int = 2,
     ):
         super(UNet, self).__init__()
         self.name = "unet"
 
+        if residual and in_channels != out_channels:  # pragma: no cover
+            raise UserWarning(
```
Comment (Andrewwango) [inline deepinv/models/unet.py]: Use warnings.warn

File: deepinv/models/utils.py [hunk 8]
```diff
@@ -78,6 +81,52 @@ def test_onesplit(model, L, refield=32, sf=1):
     return E
 
 
+# Code to build 2D or 3D variants of network components, based on the 'dim' keyword
+
+
+def fix_dim(dim: str | int) -> int:
```
Comment (Andrewwango) [inline deepinv/models/utils.py:87]: Please can you add a docstring - fix_dim is a bit vague

File: deepinv/models/utils.py [hunk 9]
```diff
@@ -78,6 +81,52 @@ def test_onesplit(model, L, refield=32, sf=1):
...
+        raise ValueError(f"dim must be of type str or int, got {dim} ({type(dim)})")
+    if dim_int not in [2, 3]:  # pragma: no cover
+        raise ValueError(
+            f"Only 2D or 3D architectures are supported, got dim={dim_int}"
+        )
+    return dim_int
+
+
+def conv_nd(dim: int) -> nn.Module:
+    return {2: nn.Conv2d, 3: nn.Conv3d}[dim]
```
Comment (Andrewwango) [inline deepinv/models/utils.py:111]: This is clever Python !

File: deepinv/tests/test_adversarial.py (hunk 3 above)
Comment (Andrewwango) [inline deepinv/tests/test_adversarial.py:136]: Thanks for these Vicky

File: deepinv/tests/test_models.py (hunk 4 above)
Comment (nucli-vicky) [reply deepinv/tests/test_models.py:533]: Partially agree. Not all models support such shapes I think (e.g. the base UNet?). The main problem is that `test_onesplit` is not (yet) supported for 3D. Overall I do not like using such shapes myself so I'm a bit biased 😅

File: deepinv/tests/test_models.py (hunk 5 above)
Comment (nucli-vicky) [reply deepinv/tests/test_models.py]: Yes and no, such tests can also be used for any image-to-image network (provided conditioning is mocked when None or the net does not have any conditioning). But happy to change it to `test_3d_denoisers` !

File: deepinv/models/utils.py (hunk 8 above)
Comment (nucli-vicky) [reply deepinv/models/utils.py:87]: Agree. Do you think the function name itself also should be made more clear? I could not come up with better naming for it myself.

File: deepinv/tests/test_models.py (hunk 6 above)
Comment (nucli-vicky) [reply deepinv/tests/test_models.py]: Agree, I'll see where to place it

Comment (Andrewwango) [reply deepinv/tests/test_models.py]: Just checking this is still TODO

**Comment by Andrewwango** [comment]
Just one unaddressed comment then good to go!

**Comment by nucli-vicky** [comment]
Thanks @Andrewwango , should be fixed now. I think the userguide is fine like that, but it may be a bit confusing because we have both the column "Tensor Size" and "3D variant"

**Comment by nucli-vicky** [comment]
Thanks @Andrewwango ! Don't think the failing test is related to the PR; docs test just needs a rerun (`File exists: 'ckpts/denoising/25-10-29-14:36:53'`). Should we add this to issues since such a stochastic fail can be avoided by not using the datetime?
